nbgrader_to_canvas/upload_grades.py

```diff
@@ -0,0 +1,379 @@
+        '''
+        self.assignment_status.status = 'Fetching Students'
+        canvas_students = {}
+        canvas_users = self._course.get_users()  
+        counter=0
+        for user in canvas_users:
+            if hasattr(user, "login_id") and user.login_id is not None:
+                canvas_students[user.login_id]=user.id  
+                if counter < self._num_students:
+                    counter += 1
+                    update_status(self._form_nb_assign_name, self._course_id, completion=5*counter/self._num_students)
+        return canvas_students
+    
+    def _get_student_grades(self):
+        '''Returns a dict of {id: {'posted_grade':score}} for all students in nb gradebook'''
+        update_status(self._form_nb_assign_name, self._course_id, status='Fetching Grades')
+        with Gradebook("sqlite:////mnt/nbgrader/"+self._nbgrader_course+"/grader/gradebook.db") as gb:
+            nb_assignment = gb.find_assignment(self._form_nb_assign_name)
+            
+            nb_students = gb.students            
+            nb_grade_data = {}
+
+            counter = 0            
+            for nb_student in nb_students:
+
+                nb_student_and_score = {} 
+
+                # Try to find the submission in the nbgrader database. If it doesn't exist, the
+                # MissingEntry exception will be raised and we assign them a score of None.
+                
+                try:
+                    nb_submission = gb.find_submission(nb_assignment.name, nb_student.id)
+
+                    if nb_submission.total_seconds_late > 0:
+                        nb_student_and_score['posted_grade'] = nb_submission.score*(100-int(self._late_penalty))/100
+                    else:
+                        nb_student_and_score['posted_grade'] = nb_submission.score
+
+                    if nb_student.id in self.nbgrader_feedback:
+                        nb_student_and_score['file_ids'] = [self.nbgrader_feedback[nb_student.id]]
+                        nb_student_and_score['text_comment'] = 'Transferred from datahub.'
+
+                except MissingEntry:
+                    nb_student_and_score['posted_grade'] = ''
+
+
+                # student.id will give us student's username, ie shrakibullah. we will need to compare this to
+                # canvas's login_id instead of user_id              
+
+                # convert nbgrader username to canvas id (integer)
+                canvas_student_id = self.canvas_students[nb_student.id]
+                nb_grade_data[canvas_student_id] = nb_student_and_score
+                counter += 1
+                update_status(self._form_nb_assign_name, self._course_id, completion=35+30*counter/self._num_students)
+            return nb_grade_data
+
+    def _get_assignment(self):
+        assignment = self._course.get_assignment(self._form_canvas_assign_id)
+        if assignment is None:
+            raise Exception('Invalid form canvas_assign_id')
```

Comment:
should this exception include the actual canvas assignment ID from the form to help with logging/debugging?